src/components/linkForm.jsx

```diff
@@ -0,0 +1,197 @@
+// @flow
+import React, { Component, Fragment } from "react";
+import Joi from "joi-browser";
+import Tags from "./tags";
+import Input from "./input";
+import Button from "./button";
+import Form from "./form";
+import { type LinkCreate, type Link } from "../types";
+
+type LinkData = LinkCreate & { tag: string };
+type State = {
+  linkData: LinkData,
+  errors: {
+    title: string,
+    url: string,
+    description: string,
+    tag: string
+  }
+};
+type Props = {
+  linkData?: Link,
+  onSubmit: (linkData: LinkCreate) => void,
+  loading: boolean
+};
+class LinkForm extends Component<Props, State> {
+  schema = {
+    title: Joi.string()
+      .required()
+      .label("Title"),
+    url: Joi.string()
+      .uri({ scheme: ["http", "https"] })
+      .label("Link"),
+    description: Joi.string()
+      .required()
+      .label("Description"),
+    tag: Joi.string()
+      .allow("")
+      .label("Tag"),
+    tags: Joi.array()
+  };
+
+  constructor(props: Props) {
+    super(props);
+    if (this.props.linkData) {
+      const link = { ...this.props.linkData, tag: "" };
+      delete link.shortLink;
+      delete link.id;
+      delete link.passage;
+      this.state = {
+        linkData: link,
+        errors: {
+          title: "",
+          url: "",
+          description: "",
+          tag: ""
+        }
+      };
+    } else {
+      this.state = {
+        linkData: {
+          url: "",
+          description: "",
+          tag: "",
+          title: "",
+          tags: []
+        },
+        errors: {
+          title: "",
+          url: "",
+          description: "",
+          tag: ""
+        }
+      };
+    }
+  }
+
+  validate = () => {
+    const errors = {
+      title: "",
+      url: "",
+      description: "",
+      tag: ""
+    };
+    const { error } = Joi.validate(this.state.linkData, this.schema, {
+      abortEarly: false
+    });
+    if (!error) {
+      return errors;
+    }
+    error.details.forEach(item => {
+      errors[item.path[0]] = item.message;
+    });
+    return errors;
+  };
+
+  handleSubmit = (e: SyntheticEvent<HTMLFormElement>) => {
+    e.preventDefault();
+    const errors = this.validate();
+    this.setState({ errors });
+    if (Object.values(errors).every(error => !error)) {
+      this.props.onSubmit(this.state.linkData);
+    }
+  };
+
+  addTag = (linkData: LinkData) => {
+    const { tag, tags } = linkData;
+    if (tag[tag.length - 1] === " ") {
+      const newTag = tag.replace(/\s/g, "");
+      if (tags.indexOf(newTag) === -1) {
+        tags.push(newTag);
+      }
+      return { ...linkData, tags, tag: " " };
+    }
+    return linkData;
+  };
+
+  deleteTag = (tag: string) => {
+    this.setState(prevState => {
+      const linkData = { ...prevState.linkData };
+      const index = linkData.tags.indexOf(tag);
```

Comment:
State mutation!
The array hasn't been re-created!
To avoid state mutation you should re-create array before splicing it.